### PR TITLE
Sir 252 deposits

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            DepositState = depositState;
+            Sources = sources;
+            Address = address;
+            DetectedDateTime = detectedDateTime;
+            ConfirmedDateTime = confirmedDateTime;
+            CompletedDateTime = completedDateTime;
+            FailedDateTime = failedDateTime;
+            CancelledDateTime = cancelledDateTime;
+
+            Events.Add(new DepositUpdated()
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:50]: This should be in the Create method. We don't need to produce events on the restoring the object from the DB

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+        public List<object> Events { get; } = new List<object>();
+
+        public static Deposit Create(
+            long id,
+            long brokerAccountRequisitesId,
+            long? accountRequisitesId,
+            long assetId,
+            decimal amount,
+            IReadOnlyCollection<DepositFee> fees,
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:119]: There are no fees associated with deposit when it's being created. We are not interested in the fees spent on the deposit tx. Fees of the Deposit object are fees spent on deposit consolidation tx and deposit provisioning tx.

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            TransactionInfo transactionInfo,
+            DepositError error,
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:121]: It is impossible to create a deposit with error

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        public long BrokerAccountRequisitesId { get; }
+        public long? AccountRequisitesId { get; }
+        public long AssetId { get; }
+        public decimal Amount { get; }
+        public IReadOnlyCollection<DepositFee> Fees { get; }
+        public TransactionInfo TransactionInfo { get; }
+        public DepositError Error { get; }
+        public DepositState DepositState { get; }
+        public IReadOnlyCollection<DepositSource> Sources { get; }
+        public string Address { get; }
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:104]: There is no `Address` property in the Deposit

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                address,
+                detectedDateTime,
+                confirmedDateTime,
+                completedDateTime,
+                failedDateTime,
+                cancelledDateTime);
+        }
+    }
+
+    public class DepositFee
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:188]: Move all these types to the separate files

Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs:188]: And we can have just a fee (for deposits, withdrawals) in Brokerage

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        {
+            Address = address;
+            Amount = amount;
+        }
+        public string Address { get; set; }
+
+        public decimal Amount { get; set; }
+    }
+
+    public class TransactionInfo
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:213]: It's shared between deposits and withdrawals

File: src/Brokerage.Common/Domain/Deposits/DepositError.cs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+﻿namespace Brokerage.Common.Domain.Deposits
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/DepositError.cs:1]: empty file

File: src/Brokerage.Common/Domain/Deposits/DepositsDetector.cs [hunk 8]
```diff
@@ -125,15 +154,67 @@ public async Task Detect(TransactionDetected transaction)
...
+                    foreach (var evt in deposit.Events)
+                    {
+                        await _publishEndpoint.Publish(evt);
+                    }
+                }
+                //Catch optimistic concurrency exception
+                catch (Exception e)
+                {
+                    //Get
+                    deposit = await _depositsRepository.GetOrDefaultAsync(
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/DepositsDetector.cs:210]: We don't need it here. Reliability concern will be resolved later on in a proper way

File: src/Brokerage.MessagingContract/DepositUpdated.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    public class DepositTransactionInfo
+    {
+        public string TransactionId { get; set; }
+        public long TransactionBlock { get; set; }
+        public long RequiredConfirmationsCount { get; set; }
+        public DateTime DateTime { get; set; }
+    }
+
+    public class DepositFee
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.MessagingContract/DepositUpdated.cs:35]: It's just a fee (will be used in withdrawals too) see Miro

File: src/Brokerage.MessagingContract/DepositUpdated.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+        public DepositState State { get; set; }
+        public DateTime DetectedDateTime { get; set; }
+        public DateTime? ConfirmedDateTime { get; set; }
+        public DateTime? CompletedDateTime { get; set; }
+        public DateTime? FailedDateTime { get; set; }
+        public DateTime? CancelledDateTime { get; set; }
+    }
+
+    public class DepositTransactionInfo
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.MessagingContract/DepositUpdated.cs:27]: The same as the fee bellow

File: src/Brokerage.MessagingContract/DepositUpdated.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        public DateTime DateTime { get; set; }
+    }
+
+    public class DepositFee
+    {
+        public long AssetId { get; set; }
+        public decimal Amount { get; set; }
+    }
+
+    public enum DepositState
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.MessagingContract/DepositUpdated.cs:41]: Move all these files to the separate files

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 1 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs:50]: fixed

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 2 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs:119]: fixed

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 3 above)
Comment (OttoVT) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs:121]: fixed

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 1 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs]: Not fixed yet

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs [hunk 12]
```diff
@@ -130,12 +85,11 @@ public class Deposit
                 accountRequisitesId,
                 assetId,
                 amount,
-                fees,
+                new DepositFee[0],
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Domain/Deposits/Deposit.cs:88]: Array.Empty<DepositFee>

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 1 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs]: This should happen only when `Create` method is called, but not `Restore`

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 5 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs]: This not fixed yet. Not `DepositFee`, but `Fee` and this model related to both deposits and withdrawals. See Miro

File: src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs [hunk 13]
```diff
@@ -54,10 +54,20 @@ private static void BuildDepositsEntity(ModelBuilder modelBuilder)
...
+                     x.TransferId
+                });
 
             modelBuilder.Entity<DepositFeeEntity>()
-                .ToTable(Tables.DepositFees);
+                .ToTable(Tables.DepositFees)
+                .HasKey(x => new
+                {
+                    x.DepositId,
+                    x.TransferId
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs:69]: I would say `DepositId` + `AssetId`

File: src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs [hunk 14]
```diff
@@ -54,10 +54,20 @@ private static void BuildDepositsEntity(ModelBuilder modelBuilder)
                 .HasKey(x => x.Id);
 
             modelBuilder.Entity<DepositSourceEntity>()
-                .ToTable(Tables.DepositSources);
+                .ToTable(Tables.DepositSources)
+                .HasKey(x => new
+                {
+                     x.DepositId,
+                     x.TransferId
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs:61]: I would say `DepositId` and `Address`

File: src/Brokerage.Common/Persistence/Entities/Deposits/DepositFeeEntity.cs [hunk 15]
```diff
@@ -7,6 +7,8 @@ public class DepositFeeEntity
     {
         public long DepositId { get; set; }
 
+        public long TransferId { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Entities/Deposits/DepositFeeEntity.cs:10]: Don't needed

File: src/Brokerage.Common/Persistence/Entities/Deposits/DepositSourceEntity.cs [hunk 16]
```diff
@@ -6,6 +6,8 @@ namespace Brokerage.Common.Persistence.Entities.Deposits
     public class DepositSourceEntity
     {
         public long DepositId { get; set; }
+        
+        public long TransferId { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.Common/Persistence/Entities/Deposits/DepositSourceEntity.cs:10]: Don't needed

File: src/Brokerage.MessagingContract/DepositTransactionInfo.cs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Swisschain.Sirius.Brokerage.MessagingContract
+{
+    public class DepositTransactionInfo
```
Comment (KonstantinRyazantsev) [inline src/Brokerage.MessagingContract/DepositTransactionInfo.cs:5]: Just a TransactionInfo

File: src/Brokerage.MessagingContract/DepositUpdated.cs (hunk 9 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.MessagingContract/DepositUpdated.cs]: Not fixed yet

File: src/Brokerage.MessagingContract/DepositUpdated.cs (hunk 10 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.MessagingContract/DepositUpdated.cs]: Not fixed yet in the domain models

File: src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs (hunk 13 above)
Comment (OttoVT) [reply src/Brokerage.Common/Persistence/DbContexts/DatabaseContext.cs:69]: nope

File: src/Brokerage.Common/Domain/Deposits/Deposit.cs (hunk 1 above)
Comment (KonstantinRyazantsev) [reply src/Brokerage.Common/Domain/Deposits/Deposit.cs]: Still not resolved

**Pull request merged**